docs/design/sql/08-expressions.md

```diff
@@ -0,0 +1,192 @@
+filters (`SELECT * FROM t WHERE a + b < 10`). Expressions may be roughly
+subdivided into 4 general categories:
+
+- Terminal expressions not referring to other expressions: column,
+  literal, parameter and type expressions.
+- Operators, like numeric arithmetic operators: `+`, `-`, etc.
+- Predicates, like comparison predicates: `<`, `>`, etc.
+- Functions, like `ABS`, `SIN`, etc.
+
+## Implementation Details
+
+We use facilities provided by Calcite to support expression parsing and
+validation. Every SQL expression goes through the following stages:
+
+- Parsing
+- Validation
+- Conversion
+- Translation
+- Evaluation
+
+### Parsing Stage
+
+The goal of the parsing stage is to obtain an AST expression
+representation from the textual expression representation for
+syntactically valid constructs and report errors for invalid ones. For
+instance, `a + b` is syntactically valid and `a +` is invalid.
+
+In Calcite, AST nodes are represented as subclasses of `SqlNode`. Column
+and type references are represented by `SqlIdentifier`. Literals are
+represented by `SqlLiteral` and its subclasses. Parameters are
+represented by `SqlDynamicParam`. Most other, if not all, expressions
+are subclasses of `SqlCall`.
+
+### Validation Stage
+
+The goal of the validation stage is to make sure that the AST
+expressions produced by the previous parsing stage are semantically
+valid. For instance, `1 + 1` is semantically valid while `1 + true` is
+not.
+
+In Calcite, the validation logic is encapsulated and orchestrated by
+implementations of `SqlValidator`. Specifically, `SqlValidatorImpl`
+extended by `HazelcastSqlValidator` is used to customize the validation
+logic to align it with our requirements.
+
+As a first step of the validation process, `SqlValidatorImpl` may apply
+some transformations to the passed AST which should not affect the
+semantics of the AST. Such transformations are called "unconditional
+rewrites" (see `SqlValidatorImpl.performUnconditionalRewrites`). Every
+`SqlCall` node has an associated `SqlOperator` returned by its
+`getOperator` method. `SqlOperator` implementations may customize
+various aspects of a certain operator including the rewriting behavior
+itself (see `SqlOperator.rewriteCall`).
+
+Normally, Calcite's parser inspects the hardcoded `SqlStdOperatorTable`
+implementation of `SqlOperatorTable` to resolve a certain operator, so
+it's hard to customize the pre-existing operators. To make the
+customization possible, operators from `SqlStdOperatorTable` are patched
+to the ones from `HazelcastSqlOperatorTable` during the AST rewriting
+process (see `HazelcastSqlValidator.performUnconditionalRewrites` and
+`HazelcastOperatorTableVisitor`).
+
+Usually, semantic validation requires assigning types to every
+expression and its operand expressions, if any. In Calcite, that type
+assignment behavior may be customized in several ways:
+
+- `SqlValidator.deriveType` may customize the behavior globally.
```

Comment:
Can we have a paragraph that explains, how all these pieces interact with each other? Namely:
- `SqlValidator.deriveType` - when is it called and how does it relate to `SqlOperator.deriveType`?
- `SqlOperandTypeInference` - when does it come into play?
- `SqlOperandTypeChecker` - same as above, and how it relates to `SqlOperator.checkOperandTypes` and `SqlOperator.getOperandCountRange` methods that have overrides in out classes (e.g. why `HazelcastSqlCastFunction` overrides `checkOperandTypes`?)

Reply:
clarified in d1d2929413835f7a462bc88686630dd57203b27c